src/cio_websocket.c

```diff
@@ -539,24 +600,39 @@ static void get_first_length(struct cio_buffered_stream *bs, void *handler_conte
 	if (field <= CIO_WEBSOCKET_SMALL_FRAME_SIZE) {
 		ws->read_frame_length = (uint64_t)field;
 		get_mask_or_payload(ws, bs, buffer);
-	} else if (field == CIO_WEBSOCKET_SMALL_FRAME_SIZE + 1) {
-		bs->read_exactly(bs, buffer, 2, get_length16, ws);
 	} else {
-		bs->read_exactly(bs, buffer, 8, get_length64, ws);
+		if (unlikely(is_control_frame(ws->ws_flags.opcode))) {
+			handle_error(ws, CIO_WEBSOCKET_CLOSE_PROTOCOL_ERROR, "payload of control frame too long");
+			return;
+		}
+
+		if (field == CIO_WEBSOCKET_SMALL_FRAME_SIZE + 1) {
+			err = bs->read_exactly(bs, buffer, 2, get_length16, ws);
+		} else {
+			err = bs->read_exactly(bs, buffer, 8, get_length64, ws);
+		}
+	}
+
+	if (unlikely(err != CIO_SUCCESS)) {
+		handle_error(ws, CIO_WEBSOCKET_CLOSE_INTERNAL_ERROR, "error while start reading extended websocket frame length");
 	}
 }
 
 static void get_header(struct cio_buffered_stream *bs, void *handler_context, enum cio_error err, struct cio_read_buffer *buffer)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 29 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS3776)